Error: The non-abstract class 'TodoEntry' is missing implementations for these members

It worked good before I upgrade **moor** and **moor_ffi** to the latest version.
I think this is caused by the different APIs among these version. Can you update this project to support the latest version of **moor** and **moor_ffi**, since I'd like to upgrade these packages to up-to-date ones. Thanks.

![image](https://user-images.githubusercontent.com/1920873/85495313-d81f1580-b60c-11ea-93f5-46b2b401658b.png)


**Stack trace as follow:**

Launching lib/main.dart on iPad (7th generation) in debug mode...
Running pod install...
Running Xcode build...
Xcode build done.                                           15.0s
Failed to build iOS app
Could not build the application for the simulator.
Error launching application on iPad (7th generation).
Error output from Xcode build:
↳
** BUILD FAILED **


Xcode's output:
↳
lib/data/database/database.g.dart:10:7: Error: The non-abstract class 'TodoEntry' is missing implementations for these members:
 - Insertable.toColumns
Try to either
 - provide an implementation,
 - inherit an implementation from a superclass or mixin,
 - mark the class as abstract, or
 - provide a 'noSuchMethod' implementation.

class TodoEntry extends DataClass implements Insertable<TodoEntry> {
      ^^^^^^^^^
../../development/flutter/.pub-cache/hosted/pub.dartlang.org/moor-3.1.0/lib/src/runtime/data_class.dart:22:27: Context: 'Insertable.toColumns' is defined here.
  Map<String, Expression> toColumns(bool nullToAbsent);
                          ^^^^^^^^^
lib/data/database/database.g.dart:104:7: Error: The non-abstract class 'TodosCompanion' is missing implementations for these members:
 - Insertable.toColumns
Try to either
 - provide an implementation,
 - inherit an implementation from a superclass or mixin,
 - mark the class as abstract, or
 - provide a 'noSuchMethod' implementation.

class TodosCompanion extends UpdateCompanion<TodoEntry> {
      ^^^^^^^^^^^^^^
../../development/flutter/.pub-cache/hosted/pub.dartlang.org/moor-3.1.0/lib/src/runtime/data_class.dart:22:27: Context: 'Insertable.toColumns' is defined here.
  Map<String, Expression> toColumns(bool nullToAbsent);
                          ^^^^^^^^^
lib/data/database/database.g.dart:252:7: Error: The non-abstract class 'Category' is missing implementations for these members:
 - Insertable.toColumns
Try to either
 - provide an implementation,
 - inherit an implementation from a superclass or mixin,
 - mark the class as abstract, or
 - provide a 'noSuchMethod' implementation.

class Category extends DataClass implements Insertable<Category> {
      ^^^^^^^^
../../development/flutter/.pub-cache/hosted/pub.dartlang.org/moor-3.1.0/lib/src/runtime/data_class.dart:22:27: Context: 'Insertable.toColumns' is defined here.
  Map<String, Expression> toColumns(bool nullToAbsent);
                          ^^^^^^^^^
lib/data/database/database.g.dart:316:7: Error: The non-abstract class 'CategoriesCompanion' is missing implementations for these members:
 - Insertable.toColumns
Try to either
 - provide an implementation,
 - inherit an implementation from a superclass or mixin,
 - mark the class as abstract, or
 - provide a 'noSuchMethod' implementation.

class CategoriesCompanion extends UpdateCompanion<Category> {
      ^^^^^^^^^^^^^^^^^^^
../../development/flutter/.pub-cache/hosted/pub.dartlang.org/moor-3.1.0/lib/src/runtime/data_class.dart:22:27: Context: 'Insertable.toColumns' is defined here.
  Map<String, Expression> toColumns(bool nullToAbsent);
                          ^^^^^^^^^
lib/data/database/database.g.dart:191:56: Error: The parameter 'd' of the method '$TodosTable.validateIntegrity' has type 'TodosCompanion', which does not match the corresponding type, 'Insertable<TodoEntry>', in the overridden method, 'TableInfo.validateIntegrity'.
 - 'TodosCompanion' is from 'package:flutterapp2/data/database/database.dart' ('lib/data/database/database.dart').
 - 'Insertable' is from 'package:moor/src/runtime/data_class.dart' ('../../development/flutter/.pub-cache/hosted/pub.dartlang.org/moor-3.1.0/lib/src/runtime/data_class.dart').
 - 'TodoEntry' is from 'package:flutterapp2/data/database/database.dart' ('lib/data/database/database.dart').
Change to a supertype of 'Insertable<TodoEntry>', or, for a covariant parameter, a subtype.
  VerificationContext validateIntegrity(TodosCompanion d,
                                                       ^
../../development/flutter/.pub-cache/hosted/pub.dartlang.org/moor-3.1.0/lib/src/runtime/query_builder/schema/table_info.dart:49:23: Context: This is the overridden method ('validateIntegrity').
  VerificationContext validateIntegrity(Insertable<D> instance,
                      ^
lib/data/database/database.g.dart:372:61: Error: The parameter 'd' of the method '$CategoriesTable.validateIntegrity' has type 'CategoriesCompanion', which does not match the corresponding type, 'Insertable<Category>', in the overridden method, 'TableInfo.validateIntegrity'.
 - 'CategoriesCompanion' is from 'package:flutterapp2/data/database/database.dart' ('lib/data/database/database.dart').
 - 'Insertable' is from 'package:moor/src/runtime/data_class.dart' ('../../development/flutter/.pub-cache/hosted/pub.dartlang.org/moor-3.1.0/lib/src/runtime/data_class.dart').
 - 'Category' is from 'package:flutterapp2/data/database/database.dart' ('lib/data/database/database.dart').
Change to a supertype of 'Insertable<Category>', or, for a covariant parameter, a subtype.
  VerificationContext validateIntegrity(CategoriesCompanion d,
                                                            ^
../../development/flutter/.pub-cache/hosted/pub.dartlang.org/moor-3.1.0/lib/src/runtime/query_builder/schema/table_info.dart:49:23: Context: This is the overridden method ('validateIntegrity').
  VerificationContext validateIntegrity(Insertable<D> instance,
                      ^
lib/data/database/database.g.dart:232:19: Error: Expected 1 type arguments.
      map['id'] = Variable<int, IntType>(d.id.value);
                  ^
../../development/flutter/.pub-cache/hosted/pub.dartlang.org/moor-3.1.0/lib/src/runtime/query_builder/expressions/variables.dart:22:9: Context: Found this candidate, but the arguments don't match.
  const Variable(this.value);
        ^^^^^^^^
lib/data/database/database.g.dart:235:24: Error: Expected 1 type arguments.
      map['content'] = Variable<String, StringType>(d.content.value);
                       ^
../../development/flutter/.pub-cache/hosted/pub.dartlang.org/moor-3.1.0/lib/src/runtime/query_builder/expressions/variables.dart:22:9: Context: Found this candidate, but the arguments don't match.
  const Variable(this.value);
        ^^^^^^^^
lib/data/database/database.g.dart:238:28: Error: Expected 1 type arguments.
      map['target_date'] = Variable<DateTime, DateTimeType>(d.targetDate.value);
                           ^
../../development/flutter/.pub-cache/hosted/pub.dartlang.org/moor-3.1.0/lib/src/runtime/query_builder/expressions/variables.dart:22:9: Context: Found this candidate, but the arguments don't match.
  const Variable(this.value);
        ^^^^^^^^
lib/data/database/database.g.dart:241:25: Error: Expected 1 type arguments.
      map['category'] = Variable<int, IntType>(d.category.value);
                        ^
../../development/flutter/.pub-cache/hosted/pub.dartlang.org/moor-3.1.0/lib/src/runtime/query_builder/expressions/variables.dart:22:9: Context: Found this candidate, but the arguments don't match.
  const Variable(this.value);
        ^^^^^^^^
lib/data/database/database.g.dart:401:19: Error: Expected 1 type arguments.
      map['id'] = Variable<int, IntType>(d.id.value);
                  ^
../../development/flutter/.pub-cache/hosted/pub.dartlang.org/moor-3.1.0/lib/src/runtime/query_builder/expressions/variables.dart:22:9: Context: Found this candidate, but the arguments don't match.
  const Variable(this.value);
        ^^^^^^^^
lib/data/database/database.g.dart:404:21: Error: Expected 1 type arguments.
      map['desc'] = Variable<String, StringType>(d.description.value);
                    ^
../../development/flutter/.pub-cache/hosted/pub.dartlang.org/moor-3.1.0/lib/src/runtime/query_builder/expressions/variables.dart:22:9: Context: Found this candidate, but the arguments don't match.
  const Variable(this.value);
        ^^^^^^^^

Command PhaseScriptExecution failed with a nonzero exit code
note: Using new build system
note: Building targets in parallel
note: Planning build
note: Constructing build description

--- a/lib/data/database/database.g.dart
+++ b/lib/data/database/database.g.dart
@@ -33,28 +33,25 @@ class TodoEntry extends DataClass implements Insertable<TodoEntry> {
           intType.mapFromDatabaseResponse(data['${effectivePrefix}category']),
     );
   }
-  factory TodoEntry.fromJson(Map<String, dynamic> json,
-      {ValueSerializer serializer = const ValueSerializer.defaults()}) {
-    return TodoEntry(
-      id: serializer.fromJson<int>(json['id']),
-      content: serializer.fromJson<String>(json['content']),
-      targetDate: serializer.fromJson<DateTime>(json['targetDate']),
-      category: serializer.fromJson<int>(json['category']),
-    );
-  }
   @override
-  Map<String, dynamic> toJson(
-      {ValueSerializer serializer = const ValueSerializer.defaults()}) {
-    return {
-      'id': serializer.toJson<int>(id),
-      'content': serializer.toJson<String>(content),
-      'targetDate': serializer.toJson<DateTime>(targetDate),
-      'category': serializer.toJson<int>(category),
-    };
+  Map<String, Expression> toColumns(bool nullToAbsent) {
+    final map = <String, Expression>{};
+    if (!nullToAbsent || id != null) {
+      map['id'] = Variable<int>(id);
+    }
+    if (!nullToAbsent || content != null) {
+      map['content'] = Variable<String>(content);
+    }
+    if (!nullToAbsent || targetDate != null) {
+      map['target_date'] = Variable<DateTime>(targetDate);
+    }
+    if (!nullToAbsent || category != null) {
+      map['category'] = Variable<int>(category);
+    }
+    return map;
   }
 
-  @override
-  TodosCompanion createCompanion(bool nullToAbsent) {
+  TodosCompanion toCompanion(bool nullToAbsent) {
     return TodosCompanion(
       id: id == null && nullToAbsent ? const Value.absent() : Value(id),
       content: content == null && nullToAbsent
@@ -67,6 +64,27 @@ class TodoEntry extends DataClass implements Insertable<TodoEntry> {
           ? const Value.absent()
           : Value(category),
     );
+  }
+
+  factory TodoEntry.fromJson(Map<String, dynamic> json,
+      {ValueSerializer serializer}) {
+    serializer ??= moorRuntimeOptions.defaultSerializer;
+    return TodoEntry(
+      id: serializer.fromJson<int>(json['id']),
+      content: serializer.fromJson<String>(json['content']),
+      targetDate: serializer.fromJson<DateTime>(json['targetDate']),
+      category: serializer.fromJson<int>(json['category']),
+    );
+  }
+  @override
+  Map<String, dynamic> toJson({ValueSerializer serializer}) {
+    serializer ??= moorRuntimeOptions.defaultSerializer;
+    return <String, dynamic>{
+      'id': serializer.toJson<int>(id),
+      'content': serializer.toJson<String>(content),
+      'targetDate': serializer.toJson<DateTime>(targetDate),
+      'category': serializer.toJson<int>(category),
+    };
   }
 
   TodoEntry copyWith(
@@ -92,7 +110,7 @@ class TodoEntry extends DataClass implements Insertable<TodoEntry> {
   int get hashCode => $mrjf($mrjc(id.hashCode,
       $mrjc(content.hashCode, $mrjc(targetDate.hashCode, category.hashCode))));
   @override
-  bool operator ==(other) =>
+  bool operator ==(dynamic other) =>
       identical(this, other) ||
       (other is TodoEntry &&
           other.id == this.id &&
@@ -118,6 +136,20 @@ class TodosCompanion extends UpdateCompanion<TodoEntry> {
     this.targetDate = const Value.absent(),
     this.category = const Value.absent(),
   }) : content = Value(content);
+  static Insertable<TodoEntry> custom({
+    Expression<int> id,
+    Expression<String> content,
+    Expression<DateTime> targetDate,
+    Expression<int> category,
+  }) {
+    return RawValuesInsertable({
+      if (id != null) 'id': id,
+      if (content != null) 'content': content,
+      if (targetDate != null) 'target_date': targetDate,
+      if (category != null) 'category': category,
+    });
+  }
+
   TodosCompanion copyWith(
       {Value<int> id,
       Value<String> content,
@@ -129,6 +161,24 @@ class TodosCompanion extends UpdateCompanion<TodoEntry> {
       targetDate: targetDate ?? this.targetDate,
       category: category ?? this.category,
     );
+  }
+
+  @override
+  Map<String, Expression> toColumns(bool nullToAbsent) {
+    final map = <String, Expression>{};
+    if (id.present) {
+      map['id'] = Variable<int>(id.value);
+    }
+    if (content.present) {
+      map['content'] = Variable<String>(content.value);
+    }
+    if (targetDate.present) {
+      map['target_date'] = Variable<DateTime>(targetDate.value);
+    }
+    if (category.present) {
+      map['category'] = Variable<int>(category.value);
+    }
+    return map;
   }
 }
 
@@ -188,31 +238,28 @@ class $TodosTable extends Todos with TableInfo<$TodosTable, TodoEntry> {
   @override
   final String actualTableName = 'todos';
   @override
-  VerificationContext validateIntegrity(TodosCompanion d,
+  VerificationContext validateIntegrity(Insertable<TodoEntry> instance,
       {bool isInserting = false}) {
     final context = VerificationContext();
-    if (d.id.present) {
-      context.handle(_idMeta, id.isAcceptableValue(d.id.value, _idMeta));
-    } else if (id.isRequired && isInserting) {
-      context.missing(_idMeta);
+    final data = instance.toColumns(true);
+    if (data.containsKey('id')) {
+      context.handle(_idMeta, id.isAcceptableOrUnknown(data['id'], _idMeta));
     }
-    if (d.content.present) {
+    if (data.containsKey('content')) {
       context.handle(_contentMeta,
-          content.isAcceptableValue(d.content.value, _contentMeta));
-    } else if (content.isRequired && isInserting) {
+          content.isAcceptableOrUnknown(data['content'], _contentMeta));
+    } else if (isInserting) {
       context.missing(_contentMeta);
     }
-    if (d.targetDate.present) {
-      context.handle(_targetDateMeta,
-          targetDate.isAcceptableValue(d.targetDate.value, _targetDateMeta));
-    } else if (targetDate.isRequired && isInserting) {
-      context.missing(_targetDateMeta);
+    if (data.containsKey('target_date')) {
+      context.handle(
+          _targetDateMeta,
+          targetDate.isAcceptableOrUnknown(
+              data['target_date'], _targetDateMeta));
     }
-    if (d.category.present) {
+    if (data.containsKey('category')) {
       context.handle(_categoryMeta,
-          category.isAcceptableValue(d.category.value, _categoryMeta));
-    } else if (category.isRequired && isInserting) {
-      context.missing(_categoryMeta);
+          category.isAcceptableOrUnknown(data['category'], _categoryMeta));
     }
     return context;
   }
@@ -223,24 +270,6 @@ class $TodosTable extends Todos with TableInfo<$TodosTable, TodoEntry> {
   TodoEntry map(Map<String, dynamic> data, {String tablePrefix}) {
     final effectivePrefix = tablePrefix != null ? '$tablePrefix.' : null;
     return TodoEntry.fromData(data, _db, prefix: effectivePrefix);
-  }
-
-  @override
-  Map<String, Variable> entityToSql(TodosCompanion d) {
-    final map = <String, Variable>{};
-    if (d.id.present) {
-      map['id'] = Variable<int, IntType>(d.id.value);
-    }
-    if (d.content.present) {
-      map['content'] = Variable<String, StringType>(d.content.value);
-    }
-    if (d.targetDate.present) {
-      map['target_date'] = Variable<DateTime, DateTimeType>(d.targetDate.value);
-    }
-    if (d.category.present) {
-      map['category'] = Variable<int, IntType>(d.category.value);
-    }
-    return map;
   }
 
   @override
@@ -264,30 +293,42 @@ class Category extends DataClass implements Insertable<Category> {
           stringType.mapFromDatabaseResponse(data['${effectivePrefix}desc']),
     );
   }
-  factory Category.fromJson(Map<String, dynamic> json,
-      {ValueSerializer serializer = const ValueSerializer.defaults()}) {
-    return Category(
-      id: serializer.fromJson<int>(json['id']),
-      description: serializer.fromJson<String>(json['description']),
-    );
-  }
   @override
-  Map<String, dynamic> toJson(
-      {ValueSerializer serializer = const ValueSerializer.defaults()}) {
-    return {
-      'id': serializer.toJson<int>(id),
-      'description': serializer.toJson<String>(description),
-    };
+  Map<String, Expression> toColumns(bool nullToAbsent) {
+    final map = <String, Expression>{};
+    if (!nullToAbsent || id != null) {
+      map['id'] = Variable<int>(id);
+    }
+    if (!nullToAbsent || description != null) {
+      map['desc'] = Variable<String>(description);
+    }
+    return map;
   }
 
-  @override
-  CategoriesCompanion createCompanion(bool nullToAbsent) {
+  CategoriesCompanion toCompanion(bool nullToAbsent) {
     return CategoriesCompanion(
       id: id == null && nullToAbsent ? const Value.absent() : Value(id),
       description: description == null && nullToAbsent
           ? const Value.absent()
           : Value(description),
     );
+  }
+
+  factory Category.fromJson(Map<String, dynamic> json,
+      {ValueSerializer serializer}) {
+    serializer ??= moorRuntimeOptions.defaultSerializer;
+    return Category(
+      id: serializer.fromJson<int>(json['id']),
+      description: serializer.fromJson<String>(json['description']),
+    );
+  }
+  @override
+  Map<String, dynamic> toJson({ValueSerializer serializer}) {
+    serializer ??= moorRuntimeOptions.defaultSerializer;
+    return <String, dynamic>{
+      'id': serializer.toJson<int>(id),
+      'description': serializer.toJson<String>(description),
+    };
   }
 
   Category copyWith({int id, String description}) => Category(
@@ -306,7 +347,7 @@ class Category extends DataClass implements Insertable<Category> {
   @override
   int get hashCode => $mrjf($mrjc(id.hashCode, description.hashCode));
   @override
-  bool operator ==(other) =>
+  bool operator ==(dynamic other) =>
       identical(this, other) ||
       (other is Category &&
           other.id == this.id &&
@@ -324,11 +365,33 @@ class CategoriesCompanion extends UpdateCompanion<Category> {
     this.id = const Value.absent(),
     @required String description,
   }) : description = Value(description);
+  static Insertable<Category> custom({
+    Expression<int> id,
+    Expression<String> description,
+  }) {
+    return RawValuesInsertable({
+      if (id != null) 'id': id,
+      if (description != null) 'desc': description,
+    });
+  }
+
   CategoriesCompanion copyWith({Value<int> id, Value<String> description}) {
     return CategoriesCompanion(
       id: id ?? this.id,
       description: description ?? this.description,
     );
+  }
+
+  @override
+  Map<String, Expression> toColumns(bool nullToAbsent) {
+    final map = <String, Expression>{};
+    if (id.present) {
+      map['id'] = Variable<int>(id.value);
+    }
+    if (description.present) {
+      map['desc'] = Variable<String>(description.value);
+    }
+    return map;
   }
 }
 
@@ -369,18 +432,17 @@ class $CategoriesTable extends Categories
   @override
   final String actualTableName = 'categories';
   @override
-  VerificationContext validateIntegrity(CategoriesCompanion d,
+  VerificationContext validateIntegrity(Insertable<Category> instance,
       {bool isInserting = false}) {
     final context = VerificationContext();
-    if (d.id.present) {
-      context.handle(_idMeta, id.isAcceptableValue(d.id.value, _idMeta));
-    } else if (id.isRequired && isInserting) {
-      context.missing(_idMeta);
+    final data = instance.toColumns(true);
+    if (data.containsKey('id')) {
+      context.handle(_idMeta, id.isAcceptableOrUnknown(data['id'], _idMeta));
     }
-    if (d.description.present) {
+    if (data.containsKey('desc')) {
       context.handle(_descriptionMeta,
-          description.isAcceptableValue(d.description.value, _descriptionMeta));
-    } else if (description.isRequired && isInserting) {
+          description.isAcceptableOrUnknown(data['desc'], _descriptionMeta));
+    } else if (isInserting) {
       context.missing(_descriptionMeta);
     }
     return context;
@@ -392,18 +454,6 @@ class $CategoriesTable extends Categories
   Category map(Map<String, dynamic> data, {String tablePrefix}) {
     final effectivePrefix = tablePrefix != null ? '$tablePrefix.' : null;
     return Category.fromData(data, _db, prefix: effectivePrefix);
-  }
-
-  @override
-  Map<String, Variable> entityToSql(CategoriesCompanion d) {
-    final map = <String, Variable>{};
-    if (d.id.present) {
-      map['id'] = Variable<int, IntType>(d.id.value);
-    }
-    if (d.description.present) {
-      map['desc'] = Variable<String, StringType>(d.description.value);
-    }
-    return map;
   }
 
   @override
@@ -423,9 +473,12 @@ abstract class _$Database extends GeneratedDatabase {
       'UPDATE todos SET category = NULL WHERE category = ?',
       variables: [Variable.withInt(var1)],
       updates: {todos},
+      updateKind: UpdateKind.update,
     );
   }
 
   @override
-  List<TableInfo> get allTables => [todos, categories];
+  Iterable<TableInfo> get allTables => allSchemaEntities.whereType<TableInfo>();
+  @override
+  List<DatabaseSchemaEntity> get allSchemaEntities => [todos, categories];
 }

--- a/pubspec.yaml
+++ b/pubspec.yaml
@@ -4,7 +4,7 @@ description: A new Flutter project.
 version: 1.0.0+1
 
 environment:
-  sdk: ">=2.5.0-dev <3.0.0"
+  sdk: ">=2.6.0 <3.0.0"
 
 dependencies:
   flutter:
@@ -12,13 +12,13 @@ dependencies:
   
   # Dart
   cupertino_icons: ^0.1.2
-  moor: ^2.1.0
+  moor: ^3.1.0
   rxdart: ^0.22.2
   intl: ^0.16.0
 
   # Database for mobile and Desktop
-  moor_ffi: ^0.2.0
-  
+  moor_ffi: ^0.6.0
+ 
   # Helper to find the database path on mobile
   path_provider: ^1.3.0
   path: ^1.6.4
@@ -26,7 +26,7 @@ dependencies:
   # Web
 
 dev_dependencies:
-  moor_generator: ^2.1.1
+  moor_generator: ^3.1.0
   build_runner: ^1.6.7
   build_web_compilers: ^2.3.0
   flutter_test:

